File: packages/react-native-client/ios/Utils/AspectRatio.swift
import FishjamCloudClient

struct AspectRatio {
    let width: Int32
    let height: Int32

    static var zero: AspectRatio {
        AspectRatio()
    }

    private init() {
        width = 0
        height = 0
    }

    init(width: Int32, height: Int32) {
        if width <= 0 || height <= 0 {
            self = .zero
        } else {
            let divisor = AspectRatio.gcd(width, height)
            self.width = width / divisor
            self.height = height / divisor
        }
    }

    init(dimensions: Dimensions?) {
        if let dimensions = dimensions {
            self = AspectRatio(width: dimensions.width, height: dimensions.height)
        } else {
            self = AspectRatio(width: 0, height: 0)
        }
    }

    func toDict() -> [String: Int32] {
        [
            "width": width,
            "height": height,
        ]
    }

    private static func gcd(_ a: Int32, _ b: Int32) -> Int32 {
        return b == 0 ? a : gcd(b, a % b)
    }
}

File: packages/react-native-client/android/src/main/java/io/fishjam/reactnative/utils/AspectRatio.kt
package io.fishjam.reactnative.utils

import com.fishjamcloud.client.models.Dimensions

data class AspectRatio(
  val width: Int,
  val height: Int
) {
  companion object {
    private val zero = AspectRatio(0, 0)

    private fun gcd(
      a: Int,
      b: Int
    ): Int = if (b == 0) a else gcd(b, a % b)

    fun create(
      width: Int,
      height: Int
    ): AspectRatio {
      if (width <= 0 || height <= 0) {
        return zero
      }
      val divisor = gcd(width, height)
      return AspectRatio(width / divisor, height / divisor)
    }

    fun create(dimensions: Dimensions?): AspectRatio =
      if (dimensions != null) {
        create(dimensions.width, dimensions.height)
      } else {
        zero
      }
  }

  fun toMap(): Map<String, Int> =
    mapOf(
      "width" to width,
      "height" to height
    )
}
